### PR TITLE
:sparkles: Adds more e2e functional tests 

### PR DESCRIPTION
**Pull request by bhargavmad**
**What this PR does / why we need it**:

1. Adds Create/Delete multiple workload clusters test scenario

- clusters in different namespaces
- clusters within same namespace

2. Adds e2e tests for delete machine lifecycle
- delete a machine using kindclient
- wait for the MachineDeployment to be reconciled

3. Adds e2e tests for multi AZ deployment
- Deploy ControlPlane and MachineDeployment in new AZ
- Deploy ControlPlane and MachineDeployment in new AZ using subnet-id

### CODE REVIEW

File: test/e2e/aws_test.go [hunk 1]
```diff
@@ -138,8 +123,87 @@ var _ = Describe("functional tests", func() {
...
+			It("should create first cluster", func() {
+				ns1 = namespace
+				clName1 = clusterName
+				By("Creating first cluster with single control plane")
+				makeSingleControlPlaneCluster(namespace, clusterName, awsClusterName, cpAWSMachinePrefix, cpBootstrapConfigPrefix, cpMachinePrefix, instanceType, testTmpDir)
+
+				By("Creating the MachineDeployment")
+				createMachineDeployment(namespace, clusterName, machineDeploymentName, awsMachineTemplateName, mdBootstrapConfig, instanceType, initialReplicas)
+			})
+			It("should create second cluster in a different namespace", func() {
```
Comment (bhargavmad) [inline test/e2e/aws_test.go]: Making a new spec keeps the code clean but on the down side, increases the num of specs which may not be acceptable. Will be happy to hear any suggestions.

File: test/e2e/aws_test.go [hunk 2]
```diff
@@ -138,8 +127,297 @@ var _ = Describe("functional tests", func() {
...
+	}
+	result, err := ec2c.DescribeInstances(instanceInput)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(len(result.Reservations)).To(Equal(int(numOfInstances)))
+}
+
+func getAvailabilityZone(ec2c *ec2.EC2) *string {
+	azs, err := ec2c.DescribeAvailabilityZones(nil)
+	Expect(err).NotTo(HaveOccurred())
+	return azs.AvailabilityZones[1].ZoneName
```
Comment (gsadhani) [inline test/e2e/aws_test.go]: Why always the second AZ? There can be some regions with only one AZ, better check the size of the AZ list before trying to access the second one.

File: test/e2e/aws_test.go [hunk 3]
```diff
@@ -138,8 +127,297 @@ var _ = Describe("functional tests", func() {
...
+}
+
+func configureAWSresources(ec2c *ec2.EC2, clusterName string, azName *string) {
```
Comment (gsadhani) [inline test/e2e/aws_test.go]: Will the resources created here get cleaned up with cluster delete?

File: test/e2e/aws_test.go (hunk 2 above)
Comment (bhargavmad) [reply test/e2e/aws_test.go]: yeah i did think of this before. But the tests are usually run on `us-west-2` so i just took the second one.

File: test/e2e/aws_test.go (hunk 3 above)
Comment (bhargavmad) [reply test/e2e/aws_test.go]: yes. they will be cleaned.

File: test/e2e/aws_test.go (hunk 1 above)
Comment (detiber) [reply test/e2e/aws_test.go]: I think the additional specs are fine here, it will give us room to optimize runtime by running them in parallel in the future.

File: test/e2e/aws_test.go (hunk 3 above)
Comment (detiber) [reply test/e2e/aws_test.go]: As discussed in the meeting the other day, we should leverage the ability to create multiple managed subnets with Cluster API here rather than manually adding additional subnets.

`AWSCluster.Spec.NetworkSpec.Subnets = [{AvailabilityZone: AZ1_NAME},{AvailabilityZone: AZ2_NAME},...]` should do the trick

Comment (bhargavmad) [reply test/e2e/aws_test.go]: Will do.

**Comment by randomvariable** [comment]
Please resolve questions if they have been addressed, thanks.

File: test/e2e/aws_test.go [hunk 4]
```diff
@@ -45,101 +47,285 @@ import (
...
-		clusterName             string
-		awsClusterName          string
-		cpMachinePrefix         string
-		cpAWSMachinePrefix      string
-		cpBootstrapConfigPrefix string
-		mdBootstrapConfig       string
-		machineDeploymentName   string
-		awsMachineTemplateName  string
-		testTmpDir              string
-		initialReplicas         int32
```
Comment (detiber) [inline test/e2e/aws_test.go]: I'd avoid this change for now to avoid potential merge conflicts with other PRs for now, it might also be good to separate this type of change as an independent PR.

File: test/e2e/aws_test.go [hunk 5]
```diff
@@ -45,101 +47,285 @@ import (
...
 	})
+
+	Describe("multiple workload clusters", func() {
+		Context("in different namespaces", func() {
+			var ns1, clName1, ns2, clName2 string
+			It("should create first cluster", func() {
+				ns1 = setup.namespace
+				clName1 = setup.clusterName
+				By("Creating first cluster with single control plane")
+				makeSingleControlPlaneCluster(setup)
```
Comment (detiber) [inline test/e2e/aws_test.go]: We seem to be doing this step in all of the different jobs now. I'm wondering if this should be moved to a BeforeEach/AfterEach instead, if we have any jobs that would require a different setup, they could be moved to a separate test file.

Not blocking for this PR, though.

File: test/e2e/aws_test.go [hunk 6]
```diff
@@ -45,101 +47,285 @@ import (
...
+
+				By("Creating the MachineDeployment")
+				createMachineDeployment(setup)
```
Comment (detiber) [inline test/e2e/aws_test.go]: Are the MachineDeployments necessary here?

File: test/e2e/aws_test.go [hunk 7]
```diff
@@ -45,101 +47,285 @@ import (
...
+			})
+			It("should create second cluster in a different namespace", func() {
+				ns2 = setup.namespace
+				clName2 = setup.clusterName
+				By("Creating second cluster with single control plane")
+				makeSingleControlPlaneCluster(setup)
+
+				By("Creating the MachineDeployment")
+				createMachineDeployment(setup)
```
Comment (detiber) [inline test/e2e/aws_test.go]: Are the MachineDeployments necessary here?

File: test/e2e/aws_test.go [hunk 8]
```diff
@@ -45,101 +47,285 @@ import (
...
+		Context("in same namespace", func() {
+			var ns, clName1, clName2 string
+			It("should create first cluster", func() {
+				ns = setup.namespace
+				clName1 = setup.clusterName
+				By("Creating first cluster with single control plane")
+				makeSingleControlPlaneCluster(setup)
+
+				By("Creating the MachineDeployment")
+				createMachineDeployment(setup)
```
Comment (detiber) [inline test/e2e/aws_test.go]: Are the MachineDeployments necessary here?

File: test/e2e/aws_test.go [hunk 9]
```diff
@@ -45,101 +47,285 @@ import (
...
+			})
+			It("should create second cluster in the same namespace", func() {
+				setup.namespace = ns
+				clName2 = setup.clusterName
+				By("Creating second cluster with single control plane")
+				makeSingleControlPlaneCluster(setup)
+
+				By("Creating the MachineDeployment")
+				createMachineDeployment(setup)
```
Comment (detiber) [inline test/e2e/aws_test.go]: Are the MachineDeployments necessary here?

File: test/e2e/aws_test.go [hunk 10]
```diff
@@ -45,101 +47,285 @@ import (
...
+
+			It("should delete both clusters", func() {
+				By("Deleting the Clusters")
+				deleteCluster(ns, clName1)
+				deleteCluster(ns, clName2)
+			})
+		})
+	})
+
+	Describe("Delete machine lifecycle", func() {
```
Comment (detiber) [inline test/e2e/aws_test.go]:
```suggestion
	Describe("MachineDeployment will replace a deleted Machine", func() {
```

File: test/e2e/aws_test.go [hunk 11]
```diff
@@ -45,101 +47,285 @@ import (
...
+			setup.availabilityZone = availabilityZones[1]
+			createAdditionalControlPlaneMachine(setup, machineName, awsMachineName, bootstrapConfigName)
+
+			By("Creating the MachineDeployment in second AZ")
+			setup.availabilityZone = availabilityZones[1]
+			createMachineDeployment(setup)
+
+			setup.subnetId = getSubnetId(privateCIDR)
+			By("Checking if the nodes are deployed in second AZ")
+			verfiyInstancesInSubnet(setup.initialReplicas+1, setup.subnetId)
```
Comment (detiber) [inline test/e2e/aws_test.go]: It would probably be better to hold off on creating any MachineDeployments until all control plane members are brought up.

File: test/e2e/aws_test.go [hunk 12]
```diff
@@ -45,101 +47,285 @@ import (
...
+
+			By("Creating the third Control Plane Machine using SubnetID in second AZ")
```
Comment (detiber) [inline test/e2e/aws_test.go]: The third control plane machine should likely be brought up in a third AZ.

File: test/e2e/aws_test.go [hunk 13]
```diff
@@ -45,101 +47,285 @@ import (
...
+			awsMachineName = setup.cpAWSMachinePrefix + "-2"
+			bootstrapConfigName = setup.cpBootstrapConfigPrefix + "-2"
+			machineName = setup.cpMachinePrefix + "-2"
+			createAdditionalControlPlaneMachine(setup, machineName, awsMachineName, bootstrapConfigName)
+
+			By("Checking if the node is deployed in second AZ")
+			verfiyInstancesInSubnet(setup.initialReplicas+2, setup.subnetId)
+
+			By("Creating the MachineDeployment using SubnetID in second AZ")
```
Comment (detiber) [inline test/e2e/aws_test.go]: Should this MachineDeployment be in a AZ that is different than the first MachineDeployment created?

File: test/e2e/aws_test.go (hunk 4 above)
Comment (bhargavmad) [reply test/e2e/aws_test.go]: Will do.

File: test/e2e/aws_test.go (hunk 5 above)
Comment (bhargavmad) [reply test/e2e/aws_test.go]: Sounds good. Will do it in a different PR

File: test/e2e/aws_test.go (hunk 6 above)
Comment (bhargavmad) [reply test/e2e/aws_test.go]: Not really. Will remove.

File: test/e2e/aws_test.go (hunk 12 above)
Comment (bhargavmad) [reply test/e2e/aws_test.go]: I was in an impression that there are commonly 2 AZs in AWS regions. Will make it 3. 👍

Comment (detiber) [reply test/e2e/aws_test.go]: Most regions have 3 azs, there are a few exceptions (some with more, some with less): https://aws.amazon.com/about-aws/global-infrastructure/regions_az/

**Comment by bhargavmad** [comment]
Rebased into a single commit
Opened a new PR to increase the timeout: https://github.com/kubernetes/test-infra/pull/15581
/test pull-cluster-api-provider-aws-e2e

**Comment by detiber** [comment]
/test pull-cluster-api-provider-aws-e2e
/lgtm
/approve
/hold
hold for another e2e test run

**Comment by detiber** [comment]
It looks like the e2e failures align with flakes in the post-submit jobs since #1348 merged, let's get this in and then we can continue to tweak the tests as needed to address any flakiness in the test signal.
/hold cancel
/lgtm
